# Assignment 3
#### Dan Cardall, Xiaoyan Mi, Pablo

In [3]:
from payoffs import *
from pricers import *

## Problem 1

In [30]:
spot = 41.0
strike = 40.0
rate = 0.08
vol = 0.3
div = 0.0
expiry = 1.0
steps = 52
reps = [1000, 10000, 25000, 50000, 75000, 100000]

In [19]:
the_call = VanillaOption(strike, expiry, call_payoff)
the_put = VanillaOption(strike, expiry, put_payoff)

In [31]:
for i in range(len(reps)):
    result_call = naive_monte_carlo_pricer(the_call, spot, rate, vol, div, reps[i])
    result_put = naive_monte_carlo_pricer(the_put, spot, rate, vol, div, reps[i]) 
    print("Naive Monte Carlo Call with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_call[0], result_call[1]))
    print("Naive Monte Carlo Put with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_put[0], result_put[1]))

Naive Monte Carlo Call with  1000  reps: 
 Estimated Price:   7.0311 
 Standard Error:   0.3412
Naive Monte Carlo Put with  1000  reps: 
 Estimated Price:   2.9624 
 Standard Error:   0.1595
Naive Monte Carlo Call with  10000  reps: 
 Estimated Price:   6.9795 
 Standard Error:   0.1064
Naive Monte Carlo Put with  10000  reps: 
 Estimated Price:   2.8966 
 Standard Error:   0.0494
Naive Monte Carlo Call with  25000  reps: 
 Estimated Price:   7.0540 
 Standard Error:   0.0682
Naive Monte Carlo Put with  25000  reps: 
 Estimated Price:   2.8923 
 Standard Error:   0.0314
Naive Monte Carlo Call with  50000  reps: 
 Estimated Price:   6.9673 
 Standard Error:   0.0478
Naive Monte Carlo Put with  50000  reps: 
 Estimated Price:   2.8846 
 Standard Error:   0.0221
Naive Monte Carlo Call with  75000  reps: 
 Estimated Price:   7.0687 
 Standard Error:   0.0395
Naive Monte Carlo Put with  75000  reps: 
 Estimated Price:   2.8767 
 Standard Error:   0.0181
Naive Monte Carlo Call with  100000  

| Number of Repitions | Estimated Price of Call | Standard Error of Call | Estimated Price of Put | Standard Error of Put|
|------|------|------|------|------|
| M = 1000   | 7.0311 | 0.3412 | 2.9624 | 0.1595 |
| M = 10000  | 6.9795 | 0.1064 | 2.8966 | 0.0494 |
| M = 25000  | 7.0540 | 0.0682 | 2.8923 | 0.0314 |
| M = 50000  | 6.9673 | 0.0478 | 2.8846 | 0.0221 |
| M = 75000  | 7.0687 | 0.0395 | 2.8767 | 0.0181 |
| M = 100000 | 6.9641 | 0.0339 | 2.8732 | 0.0156 |



### Results
We can see in the table above that as the number of repitions goes up, the standard error of both the estimated call price and the estimated put price decreases.  

## Problem 2

In [32]:
spot = 41.0
strike = 40.0
rate = 0.08
vol = 0.3
div = 0.0
expiry = 1.0
steps = 52
reps = [1000, 10000, 25000, 50000, 75000, 100000]
the_call = VanillaOption(strike, expiry, call_payoff)
the_put = VanillaOption(strike, expiry, put_payoff)

In [33]:
for i in range(len(reps)):
    result_call = antithetic_monte_carlo_pricer(the_call, spot, rate, vol, div, reps[i])
    result_put = antithetic_monte_carlo_pricer(the_put, spot, rate, vol, div, reps[i]) 
    print("Naive Monte Carlo using Antithetic Sampling Call with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_call[0], result_call[1]))
    print("Naive Monte Carlo using Antithetic Sampling Put with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_put[0], result_put[1]))

Naive Monte Carlo using Antithetic Sampling Call with  1000  reps: 
 Estimated Price:   7.0939 
 Standard Error:   0.2420
Naive Monte Carlo using Antithetic Sampling Put with  1000  reps: 
 Estimated Price:   2.9382 
 Standard Error:   0.1125
Naive Monte Carlo using Antithetic Sampling Call with  10000  reps: 
 Estimated Price:   7.0226 
 Standard Error:   0.0766
Naive Monte Carlo using Antithetic Sampling Put with  10000  reps: 
 Estimated Price:   2.9153 
 Standard Error:   0.0351
Naive Monte Carlo using Antithetic Sampling Call with  25000  reps: 
 Estimated Price:   6.9274 
 Standard Error:   0.0475
Naive Monte Carlo using Antithetic Sampling Put with  25000  reps: 
 Estimated Price:   2.8706 
 Standard Error:   0.0220
Naive Monte Carlo using Antithetic Sampling Call with  50000  reps: 
 Estimated Price:   6.9456 
 Standard Error:   0.0338
Naive Monte Carlo using Antithetic Sampling Put with  50000  reps: 
 Estimated Price:   2.8822 
 Standard Error:   0.0157
Naive Monte Carlo usin

| Number of Repitions | Estimated Price of Call | Estimated Price of Call w/ Antithetic | Standard Error of Call | Standard Error of Call w/ Antithetic |Estimated Price of Put |Estimated Price of Put w/ Antithetic | Standard Error of Put| Standard Error of Put w/ Antithetic|
|------|------|------|------|------|------|------|------|------|
| M = 1000   | 7.0311 | 7.0939 |0.3412 | 0.2420 | 2.9624 | 2.9382 |0.1595 | 0.1125 |
| M = 10000  | 6.9795 | 7.0226 |0.1064 | 0.0766 | 2.8966 | 2.9153 |0.0494 | 0.0351 |
| M = 25000  | 7.0540 | 6.9274 |0.0682 | 0.0475 | 2.8923 | 2.8706 |0.0314 | 0.0220 |
| M = 50000  | 6.9673 | 6.9456 |0.0478 | 0.0338 | 2.8846 | 2.8822 |0.0221 | 0.0157 |
| M = 75000  | 7.0687 | 6.9595 |0.0395 | 0.0276 | 2.8767 | 2.8818 |0.0181 | 0.0127 |
| M = 100000 | 6.9641 | 6.9726 |0.0339 | 0.0239 | 2.8732 | 2.8898 |0.0156 | 0.0111 |


### Results

As we can see in the table above, the standard errors using the antithetic sampling method reduces the standard error of the price estimate for the given repition size. This is true for both the estimated call and put prices. This makes sense as antithetic sampling is ultimately artificailly doubling the sample size which will decrease our standard error. 

## Problem 3

In [34]:
spot = 41.0
strike = 40.0
rate = 0.08
vol = 0.3
div = 0.0
expiry = 1.0
steps = 52
reps = [1000, 10000, 25000, 50000, 75000, 100000]
the_call = VanillaOption(strike, expiry, call_payoff)
the_put = VanillaOption(strike, expiry, put_payoff)

In [35]:
for i in range(len(reps)):
    result_call = stratified_monte_carlo_pricer(the_call, spot, rate, vol, div, reps[i])
    result_put = stratified_monte_carlo_pricer(the_put, spot, rate, vol, div, reps[i]) 
    print("Naive Monte Carlo using Stratified Sampling Call with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_call[0], result_call[1]))
    print("Naive Monte Carlo using Stratified Sampling Put with ", reps[i], " reps: \n Estimated Price: %8.4f \n Standard Error: %8.4f" % (result_put[0], result_put[1]))

Naive Monte Carlo using Stratified Sampling Call with  1000  reps: 
 Estimated Price:   6.9597 
 Standard Error:   0.3376
Naive Monte Carlo using Stratified Sampling Put with  1000  reps: 
 Estimated Price:   2.8996 
 Standard Error:   0.1591
Naive Monte Carlo using Stratified Sampling Call with  10000  reps: 
 Estimated Price:   6.9614 
 Standard Error:   0.1069
Naive Monte Carlo using Stratified Sampling Put with  10000  reps: 
 Estimated Price:   2.8869 
 Standard Error:   0.0495
Naive Monte Carlo using Stratified Sampling Call with  25000  reps: 
 Estimated Price:   6.9609 
 Standard Error:   0.0676
Naive Monte Carlo using Stratified Sampling Put with  25000  reps: 
 Estimated Price:   2.8861 
 Standard Error:   0.0313
Naive Monte Carlo using Stratified Sampling Call with  50000  reps: 
 Estimated Price:   6.9610 
 Standard Error:   0.0478
Naive Monte Carlo using Stratified Sampling Put with  50000  reps: 
 Estimated Price:   2.8859 
 Standard Error:   0.0221
Naive Monte Carlo usin

| Number of Repitions | Estimated Price of Call | Estimated Price of Call w/ Antithetic |Estimated Price of Call w/ Stratified| Standard Error of Call | Standard Error of Call w/ Antithetic |Standard Error of Call w/ Stratified|Estimated Price of Put |Estimated Price of Put w/ Antithetic |Estimated Price of Put w/ Stratified|Standard Error of Put| Standard Error of Put w/ Antithetic|Standard Error of Putw/ Stratified|
|------|------|------|------|------|------|------|------|------|------|------|------|------|
| M = 1000   | 7.0311 | 7.0939 | 6.9597 |0.3412 | 0.2420 | 0.3376 | 2.9624 | 2.9382 | 2.8996 | 0.1595 | 0.1125 | 0.1591 |
| M = 10000  | 6.9795 | 7.0226 | 6.9614 |0.1064 | 0.0766 | 0.1069 | 2.8966 | 2.9153 | 2.8869 | 0.0494 | 0.0351 | 0.0495 |
| M = 25000  | 7.0540 | 6.9274 | 6.9609 |0.0682 | 0.0475 | 0.0676 | 2.8923 | 2.8706 | 2.8861 | 0.0314 | 0.0220 | 0.0313 |
| M = 50000  | 6.9673 | 6.9456 | 6.9610 |0.0478 | 0.0338 | 0.0478 | 2.8846 | 2.8822 | 2.8859 | 0.0221 | 0.0157 | 0.0221 |
| M = 75000  | 7.0687 | 6.9595 | 6.9610 |0.0395 | 0.0276 | 0.0390 | 2.8767 | 2.8818 | 2.8858 | 0.0181 | 0.0127 | 0.0180 |
| M = 100000 | 6.9641 | 6.9726 | 6.9610 |0.0339 | 0.0239 | 0.0338 | 2.8732 | 2.8898 | 2.8858 | 0.0156 | 0.0111 | 0.0156 |


### Results

As we can see in the table above, the standard errors for the stratified samples are similar to the usual standard errors. However, through repeated samples, we noticed that the sample standard deviation from the stratified was much more consistent compared to the usual sample standard deviation. This would make sense as when we take a stratified sample, we are making the distribution of the sample much more consistent.